src/VisualStudio/Core/Test.Next/Services/LspDiagnosticsTests.cs

```diff
@@ -0,0 +1,470 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Editor.UnitTests.Workspaces;
+using Microsoft.CodeAnalysis.LanguageServer;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.LanguageServices.Implementation.LanguageService;
+using Moq;
+using Nerdbank.Streams;
+using Newtonsoft.Json.Linq;
+using Roslyn.Test.Utilities;
+using Roslyn.Utilities;
+using StreamJsonRpc;
+using Xunit;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Roslyn.VisualStudio.Next.UnitTests.Services
+{
+    [UseExportProvider]
+    public class LspDiagnosticsTests : AbstractLanguageServerProtocolTests
+    {
+        [Fact]
+        public async Task AddDiagnosticTestAsync()
+        {
+            using var workspace = CreateTestWorkspace("", out _);
+            var document = workspace.CurrentSolution.Projects.First().Documents.First();
+
+            var diagnosticsMock = new Mock<IDiagnosticService>();
+            // Create a mock that returns a diagnostic for the document.
+            SetupMockWithDiagnostics(diagnosticsMock, document.Id, await CreateMockDiagnosticDataAsync(document, "id").ConfigureAwait(false));
+
+            // Publish one document change diagnostic notification ->
+            // 1.  doc1 with id.
+            //
+            // We expect one publish diagnostic notification ->
+            // 1.  from doc1 with id.
+            var results = await RunPublishDiagnosticsAsync(workspace, diagnosticsMock.Object, 1, document).ConfigureAwait(false);
+
+            // We expect one publish notification for the diagnostic on the document.
```

Comment:
Consider `var result = Assert.Single(results)` once which both removes all the .Single()s everywhere else, but also gives a nice assert failure if it isn't right.